### DISCUSSION TITLE
Embedding designer and dashboard into 3rd party UI

### DISCUSSION THREAD

**Discussion by tomy2105** [elsa-workflows/elsa-core#800] [Q&A]
Sorry if this was discussed before.

It would be helpful if designer/dashboard would allow to be invoked without header/left navigation and/or footer in order to be easily embedded (maybe even via simple iframe) into some "larger" external UI.

Is this possible?
I'm aware I could probably use current designer code UI components but that might lead to licensing issues (https://github.com/elsa-workflows/elsa-core/discussions/798). 
And am not sure how "componentized" dashboard is and would it require external UI to be Razor/Blazor?

**Comment by sfmskywalker** [comment]
The designer can be embedded today, so that's covered.

The dashboard itself on the other hand was not designed to be embedded in a larger application. But perhaps we might consider componentizing some of those screens in addition to the workflow designer as well.
We'd have the following:

- [x] Workflow Editor
- [x] Workflow Definition List
- [x] Workflow Instance List
- [x] Workflow Registry
- [ ] Custom Activities

This way, one could easily create their own dashboard (own header. footer, main nav etc), and simply place the appropriate Elsa component on a page.

A related, important aspect is customizability - the application host should at the very least be able to override CSS classes, and ideally provide their own HTML templates for certain aspects. For instance, the `Workflow Definition List` component might expose two templates: a list template (to render e.g. `ul`/`li` instead of `tr`/`td`) and an item template (to render the contents of e.g. an `li`/`td`).

That would be quite flexible. But logical next question would be: what template parser should be used? We could of course pick e.g. handle bars, but if the dashboard happens to be a SPA built with e.g. React, it might be even sweeter to be able to use JSX. So perhaps the template parser should be pluggable as well.

Anyway, the custom templates can be figured out separate from the componentization.
Thanks for the question!

PS. I think I prematurely converted your issue to a discussion, because now I want to keep this as an issue 😅 
I'll create a new one linking to this discussion.

**Reply by sfmskywalker** [reply]
On second thought, I'm thinking that the templating thing might be necessary in order to integrate seamlessly into the overall application. So in that respect, all you'd really want is a proper Elsa API Client library to easily fetch data & issue updates. So perhaps a better component would be an embeddable "studio" that allows you to open workflow definitions from a dialog window for example. Something like what [node-red](https://nodered.org/) has.

**Comment by tomy2105** [comment]
I kind of like your first suggestion. Don't know how much work it would be....
Maybe it would be sufficient to allow embedding custom css and/or javascript inside current dashboard page. So "integrator" would be able to influence how dashboard/designer look like and or behave using those......
Do you think some "simple" ability to do something like that could make into Elsa 2.0?

**Reply by sfmskywalker** [reply]
Yep. In fact, that's exactly what I'm working on as we speak :)

**Reply by lakuri** [reply]
@sfmskywalker Did you complete it? Is it documented somewhere how we can customize the look and feel?

**Reply by sfmskywalker** [reply]
I'm afraid I never got to completing the customisability from the UI itself, unfortunately. Which means that right now, you have to include your own CSS to override styles to make the UI your own.

The plan hasn't changed though; we're gonna give it another go when we start working on our Blazor implementation of the dashboard, which will be themeable.
